psycopg/psycopg/cursor.py

```diff
@@ -156,6 +157,18 @@ def nextset(self) -> Optional[bool]:
         Return `!True` if a new result is available, which will be the one
         methods `!fetch*()` will operate on.
         """
+        # Raise a warning if people is calling nextset() in pipeline mode
+        # after a sequence of execute() in pipeline mode. Pipeline accumulating
+        # execute() results in the cursor is an unintended difference w.r.t.
+        # non-pipeline mode.
+        if self._execmany_returning is None and self._conn._pipeline:
```

Comment:
I think checking `self._conn._pipeline` is not enough because one might have exited the pipeline before calling `.nextset()`, e.g.:
~~~python
    with conn.cursor() as cur:                                                                                                          
        with conn.pipeline() as p:                                                                                                      
            cur.execute("select 1")                                                                                                     
            cur.execute("select 2")                                                                                                     
            assert len(p.result_queue) == 2                                                                                             
        (r,) = cur.fetchone()                                                                                                           
        assert r == 2  # expected after clobbering                                                                                                   
        assert cur.nextset() is None  # no warning since we're out of the pipeline         
~~~

Reply:
...I'd say if they do that they deserve an unexpected behaviour :grimacing: 

I know that a cursor can be created independently before or after the pipeline. However, it seems reasonable to require (maybe to enforce?) that if an `execute()` is performed before/inside/after a pipeline, the matching `fetch*()` should happen in the same area. I expect the unexpected, otherwise.

Is this something you would consider enforcing? An effective and efficient way to do so would be to `_reset()` the cursor upon entering and exiting the pipeline.

Reply:
I agree: calling `.fetch*()` within the same scope as respective `.execute()` is probably the most sensible things to do. So the implementation here is probably enough.

> Is this something you would consider enforcing?

Hm, no. `.fetch*()` invocations after exiting the pipeline work fine[^1] as results of pipeline's `execute()`s are retrieved at pipeline exit and bound to their cursor (so that those `.fetch*()` do not actully do I/O but just return).  So I'm not sure there is any value in changing/breaking this: that would rather be a burden on our side.

[^1]: We have tests for this: https://github.com/psycopg/psycopg/blob/915593ad6f3577f67ba73d0105fec0e88940ae8e/tests/test_pipeline.py#L131